pkg/cloudprovider/aws/launchtemplate.go

```diff
@@ -357,27 +372,49 @@ exec > >(tee /var/log/user-data.log|logger -t user-data -s 2>/dev/console) 2>&1
 			*caBundle))
 	}
 
+	clusterDNSIP := p.clusterDNSIP
+	if len(constraints.KubeletConfiguration.ClusterDNS) > 0 {
+		clusterDNSIP = net.ParseIP(constraints.KubeletConfiguration.ClusterDNS[0])
+		userData.WriteString(fmt.Sprintf(` \
+    --dns-cluster-ip '%s'`, clusterDNSIP))
+	}
+
+	ipv6Cluster := isIPv6(clusterDNSIP)
+	if ipv6Cluster {
+		userData.WriteString(` \
+--ip-family ipv6`)
+		_, ipNet, err := net.ParseCIDR(clusterDNSIP.String() + "/108")
+		if err != nil {
+			panic(fmt.Sprintf("Parsing IPv6 cluster DNS, %s", err))
+		}
+		userData.WriteString(fmt.Sprintf(` \
+--service-ipv6-cidr '%s'`, ipNet))
+	}
+
 	nodeLabelArgs := p.getNodeLabelArgs(functional.UnionStringMaps(additionalLabels, constraints.Labels))
 	nodeTaintsArgs := p.getNodeTaintArgs(constraints)
 	kubeletExtraArgs := strings.Trim(strings.Join([]string{nodeLabelArgs, nodeTaintsArgs.String()}, " "), " ")
 
-	if !injection.GetOptions(ctx).AWSENILimitedPodDensity {
+	if !injection.GetOptions(ctx).AWSENILimitedPodDensity || ipv6Cluster {
 		userData.WriteString(` \
-    --use-max-pods=false`)
+    --use-max-pods false`)
 		kubeletExtraArgs += " --max-pods=110"
 	}
 
 	if len(kubeletExtraArgs) > 0 {
 		userData.WriteString(fmt.Sprintf(` \
     --kubelet-extra-args '%s'`, kubeletExtraArgs))
-	}
-	if len(constraints.KubeletConfiguration.ClusterDNS) > 0 {
-		userData.WriteString(fmt.Sprintf(` \
-    --dns-cluster-ip '%s'`, constraints.KubeletConfiguration.ClusterDNS[0]))
 	}
 	return base64.StdEncoding.EncodeToString(userData.Bytes())
 }
 
+func isIPv6(ip net.IP) bool {
```

Comment:
It would be great if we were somehow guaranteed that this isn't nil. Then we could just use 
```
len(ip.To16()) == net.IPv6len
```
without a helper. 
Why do we even need to check To4?

Reply:
When I ran a test the (ipv4).To16() returns IPv6 formatted IPv4 address which length is IPv6len. So I added a check To4 which returns nil for IPv6 addresses.

Reply:
`net.IP` is fun and often encodes IPv4 addresses embedded in IPv6 - see my comment elsewhere for a play.golang example.

You have to check `To4() != nil`